### PR TITLE
Fixes TlsHandler from operation, perf and compat POVs.

### PR DESCRIPTION
**Pull request by nayato**
Motivation:
Fix `TlsHandler`, optimize encryption / decryption wrt Write vs Flush semantics, optimize memory retention for sparse communication, adopt `IByteBuffer.GetIoBuffers()` in `TlsHandler`.

Modifications:
- `TlsHandler` is rewritten to use batch->copy->write approach instead of writing directly to `SslStream` in order to reduce overhead of framing / writing out everything separately.
- `TlsHandler` uses `IByteBuffer.GetIoBuffers()` instead of direct array access increasing compatibility for non-array-backed buffers.
- `TlsHandler` and `TlsHandler.MediationStream` use Task semantics now when reading through `SslStream`.
- Extra: upgraded to NBench 2.2, xUnit 2.1.0.
- Extra: better portability for .NET Core support.

Result:
`TlsHandler` provides better framing and much better (x2+) performance (depends on application control of flushing).
`TlsHandler` operates properly in edge cases when more than one frame is accepted at once and if decrypted frame size is more than 4 KB.
`TlsHandler` holds only 256 byte buffer while waiting for more data to arrive off the wire.
Extra: DotNetty is almost .NET Core compatible.

### CODE REVIEW

**Comment by nayato** [comment]
@mastermanu, @mtuchkov PTAL. TlsHandler.cs is the focus (`TlsHandler.Unwrap` and `Wrap` to be precise).

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 1]
```diff
@@ -144,10 +146,10 @@ static void HandleAuthenticationCompleted(Task task, object state)
                 case TaskStatus.Faulted:
                 {
                     // ReSharper disable once AssignNullToNotNullAttribute -- task.Exception will be present as task is faulted
-                    State oldState = self.state;
-                    if ((oldState & State.AuthenticationCompleted) == 0)
+                    HandlerState oldState = self.state;
+                    if (!oldState.Has(HandlerState.AuthenticationCompleted))
```
Comment (nayato) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: `.HasAny`

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 2]
```diff
@@ -112,27 +115,26 @@ bool IgnoreException(Exception t)
...
         {
             var self = (TlsHandler)state;
             switch (task.Status)
             {
                 case TaskStatus.RanToCompletion:
                 {
-                    State oldState = self.state;
-                    if ((oldState & State.AuthenticationCompleted) == 0)
+                    HandlerState oldState = self.state;
+                    if (!oldState.Has(HandlerState.AuthenticationCompleted))
```
Comment (nayato) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: `.HasAny`

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 3]
```diff
@@ -21,23 +21,26 @@ namespace DotNetty.Handlers.Tls
 
     public sealed class TlsHandler : ByteToMessageDecoder
     {
-        const int ReadBufferSize = 4 * 1024; // todo: research perfect size
+        const int FallbackReadBufferSize = 256;
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: [Nit] add InBytes suffix please

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 4]
```diff
@@ -21,23 +21,26 @@ namespace DotNetty.Handlers.Tls
...
+        const int UnencryptedWriteBatchSize = 14 * 1024;
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: [Nit] add InBytes suffix please

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 5]
```diff
@@ -144,10 +146,10 @@ static void HandleAuthenticationCompleted(Task task, object state)
...
                     // ReSharper disable once AssignNullToNotNullAttribute -- task.Exception will be present as task is faulted
-                    State oldState = self.state;
-                    if ((oldState & State.AuthenticationCompleted) == 0)
+                    HandlerState oldState = self.state;
+                    if (!oldState.HasAny(HandlerState.AuthenticationCompleted))
                     {
-                        self.state = (oldState | State.FailedAuthentication) & ~State.Authenticating;
+                        self.state = (oldState | HandlerState.FailedAuthentication) & ~HandlerState.Authenticating;
                     }
                     self.HandleFailure(task.Exception);
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: should this only be in the above if clause? what scenario breaks if we do that?

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 6]
```diff
@@ -180,47 +182,257 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
...
+            int endOffset = input.WriterIndex;
+            int offset = startOffset;
+            int totalLength = 0;
 
-            if (!this.EnsureAuthenticated())
+            // If we calculated the length of the current SSL record before, use that information.
+            if (this.packetLength > 0)
             {
-                return;
+                if (endOffset - startOffset < this.packetLength)
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: [Nit] this isn't a method or property on IByteBuffer?

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 7]
```diff
@@ -180,47 +182,257 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
...
+            while (totalLength < TlsUtils.MAX_ENCRYPTED_PACKET_LENGTH)
+            {
+                int readableBytes = endOffset - offset;
+                if (readableBytes < TlsUtils.SSL_RECORD_HEADER_LENGTH)
+                {
+                    break;
+                }
+
+                int packetLength = TlsUtils.GetEncryptedPacketLength(input, offset);
+                if (packetLength == -1)
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: i assume its not possible to read one packet and then get -1 on the subsequent call, right?

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 8]
```diff
@@ -180,47 +182,257 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
...
+
+                if (packetLength > readableBytes)
+                {
+                    // wait until the whole packet can be read
+                    this.packetLength = packetLength;
+                    break;
+                }
+
+                int newTotalLength = totalLength + packetLength;
+                if (newTotalLength > TlsUtils.MAX_ENCRYPTED_PACKET_LENGTH)
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: so let's say we read 1 packet and then try to read the 2nd packet, but it exceeds the max_encrypted_packet_length?
don't we need to update this.packetLength? Or do we do that later in this method?

File: src/DotNetty.Handlers/Tls/TlsHandler.cs (hunk 7 above)
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: i see that we throw anyways later down the line. nevermind

---
In reply to: [64496839](https://github.com/Azure/DotNetty/pull/116#discussion_r64496839) [](ancestors = 64496839)

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 9]
```diff
@@ -180,47 +182,257 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
...
+            this.firedChannelRead = false;
+            ctx.FireChannelReadComplete();
+        }
+
+        void ReadIfNeeded(IChannelHandlerContext ctx)
+        {
+            // If handshake is not finished yet, we need more data.
+            if (!ctx.Channel.Configuration.AutoRead && (!this.firedChannelRead || !this.state.HasAny(HandlerState.AuthenticationCompleted)))
+            {
+                // No auto-read used and no message passed through the ChannelPipeline or the handhshake was not complete
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: [Nit] spelling

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 10]
```diff
@@ -619,4 +858,25 @@ public WaitHandle AsyncWaitHandle
...
+        AuthenticationCompleted = Authenticated | FailedAuthentication
+    }
+
+    static class HandlerStateExtensions
+    {
+        public static bool Has(this HandlerState value, HandlerState testValue) => (value & testValue) == testValue;
+
+        public static bool HasAny(this HandlerState value, HandlerState testValue) => (value & testValue) != 0;
+
+        public static HandlerState Sans(this HandlerState value, HandlerState exclusionValue) => value & ~exclusionValue;
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: used anywhere?

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 11]
```diff
@@ -112,27 +115,26 @@ bool IgnoreException(Exception t)
...
         {
             var self = (TlsHandler)state;
             switch (task.Status)
             {
                 case TaskStatus.RanToCompletion:
                 {
-                    State oldState = self.state;
-                    if ((oldState & State.AuthenticationCompleted) == 0)
+                    HandlerState oldState = self.state;
+                    if (!oldState.HasAny(HandlerState.AuthenticationCompleted))
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: remove if checks and replace with asserts?

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 12]
```diff
@@ -180,47 +182,257 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
...
+        void Unwrap(IChannelHandlerContext ctx, IByteBuffer packet, int offset, int length, List<object> output)
+        {
+            //bool notifyClosure = false; // todo: netty/issues/137
+            bool pending = false;
+
+            IByteBuffer decodeOut = null;
+
+            try
+            {
+                this.mediationStream.AcceptBytes(packet.GetIoBuffer(offset, length));
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: rename...AcceptBytes implies data getting *returned*

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 13]
```diff
@@ -180,47 +182,257 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
...
+
+                if (!this.EnsureAuthenticated())
+                {
+                    return;
+                }
+
+                Task<int> currentReadFuture = this.pendingReadFuture;
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: this.pendingSslStreamReadFuture

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 14]
```diff
@@ -180,47 +182,257 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
...
+
+                if (currentReadFuture != null && !currentReadFuture.IsCompleted)
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: Assert

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 15]
```diff
@@ -180,47 +182,257 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
...
+                {
+                    return;
+                }
+
+                decodeOut = this.pendingReadBuffer ?? ctx.Allocator.Buffer(length);
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: pendingSslStreamReadBuffer

Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: combine with above assert (someehow)

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 16]
```diff
@@ -180,47 +182,257 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
...
+                    {
+                        int readInstant = currentReadFuture.Result;
+
+                        Contract.Assert(readInstant > 0);
+
+                        outletOffset += readInstant;
+                        read += readInstant;
+
+                        IByteBuffer swappedOut;
+                        int remainingInput = this.mediationStream.SourceLength; // bytes not yet read by SslStream
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: rename SourceLength

File: src/DotNetty.Handlers/Tls/TlsHandler.cs (hunk 14 above)
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: add comment

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 17]
```diff
@@ -180,47 +182,257 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
...
+                            else
+                            {
+                                // current output buffer is full - very likely there's more data in SslStream's internal state
+                                decodeOut = ctx.Allocator.Buffer(Math.Max(FallbackReadBufferSize, length - read));
+                            }
+                        }
+                        else
+                        {
+                            // SslStream has more input to go through - it won't require more input on next ReadAsync call
+                            if (outlet.Count - outletOffset >= remainingInput)
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: check if decoding 2nd packet can result in 0 byte read from SSL Stream perspective. if so, need to add to output

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 18]
```diff
@@ -180,47 +182,257 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
...
+                            if (outletOffset < outlet.Count)
+                            {
+                                // current output buffer isn't full - SslStream will likely block =>
+                                // release current result and switch to buffer of FallbackReadBufferSize
+                                decodeOut = ctx.Allocator.Buffer(FallbackReadBufferSize);
+                            }
+                            else
+                            {
+                                // current output buffer is full - very likely there's more data in SslStream's internal state
+                                decodeOut = ctx.Allocator.Buffer(Math.Max(FallbackReadBufferSize, length - read));
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: why not pass remainingInput here?

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 19]
```diff
@@ -180,47 +182,257 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
...
+                            if (outlet.Count - outletOffset >= remainingInput)
+                            {
+                                // use the same output buffer
+                                swappedOut = null;
+                            }
+                            else
+                            {
+                                // release the buffer and provide another sufficient output buffer
+                                swappedOut = decodeOut;
+                                decodeOut = ctx.Allocator.Buffer(length - read);
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: remainingInput?

File: src/DotNetty.Handlers/Tls/TlsHandler.cs (hunk 18 above)
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: what's the heuristic here? technically we've exhausted all the remaining input? how much do we expect this to be?

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 20]
```diff
@@ -21,23 +21,26 @@ namespace DotNetty.Handlers.Tls
...
     public sealed class TlsHandler : ByteToMessageDecoder
     {
-        const int ReadBufferSize = 4 * 1024; // todo: research perfect size
+        const int FallbackReadBufferSize = 256;
+        const int UnencryptedWriteBatchSize = 14 * 1024;
 
         static readonly Exception ChannelClosedException = new IOException("Channel is closed");
-        static readonly Action<Task, object> AuthenticationCompletionCallback = new Action<Task, object>(HandleAuthenticationCompleted);
-        static readonly AsyncCallback SslStreamReadCallback = new AsyncCallback(HandleSslStreamRead);
+        static readonly Action<Task, object> AuthenticationCompletionCallback = new Action<Task, object>(HandleHandshakeCompleted);
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: [Nit] HandshakeCompletionCallback

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 21]
```diff
@@ -180,47 +179,306 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
...
+            int offset = startOffset;
+            int totalLength = 0;
 
-            if (!this.EnsureAuthenticated())
+            List<int> packetLengths;
+            // if we calculated the length of the current SSL record before, use that information.
+            if (this.packetLength > 0)
             {
-                return;
+                if (endOffset - startOffset < this.packetLength)
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: [Nit] add comment indicating that there is still not enough data left to read through the remaining packet.

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 22]
```diff
@@ -180,47 +179,306 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
...
+                Contract.Assert(packetLength > 0);
+
+                if (packetLength > readableBytes)
+                {
+                    // wait until the whole packet can be read
+                    this.packetLength = packetLength;
+                    break;
+                }
+
+                int newTotalLength = totalLength + packetLength;
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: btw i assume this also includes the bytes involved in encoding the length?

the comment for GetEncryptedPacketLength says that the reader index is not incremented

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 23]
```diff
@@ -180,47 +179,306 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
...
+                    packetLengths = new List<int>(4);
+                    packetLengths.Add(this.packetLength);
+                    offset += this.packetLength;
+                    totalLength = this.packetLength;
+                    this.packetLength = 0;
+                }
+            }
+            else
+            {
+                packetLengths = new List<int>(4);
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: btw, I believe this is the default length of an initialized List.

Also, can we recycle / reuse this across calls?

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 24]
```diff
@@ -180,47 +179,306 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
...
+            {
+                ArraySegment<byte> inputIoBuffer = packet.GetIoBuffer(offset, length);
+                this.mediationStream.SetSource(inputIoBuffer.Array, inputIoBuffer.Offset);
+
+                int packetIndex = 0;
+
+                while (!this.EnsureAuthenticated())
+                {
+                    this.mediationStream.ExpandSource(packetLengths[packetIndex]);
+                    if (++packetIndex >= packetLengths.Count)
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: [Nit] ==

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 25]
```diff
@@ -180,47 +179,306 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
...
+                            // we did feed the whole current packet to SslStream yet it did not produce any result -> move to the next packet in input
+                            Contract.Assert(this.mediationStream.SourceReadableBytes == 0);
+
+                            continue;
+                        }
+
+                        int read = currentReadFuture.Result;
+
+                        // Now output the result of previous read and decide whether to do an extra read on the same source or move forward
+                        AddBufferToOutput(outputBuffer, read, output);
```
Comment (mastermanu) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: [Nit] naming

File: src/DotNetty.Handlers/Tls/TlsHandler.cs (hunk 23 above)
Comment (nayato) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: default is `0`. We can see if there's visible benefit in pooling. it's different than `List<object>` that we're using in ByteToMessageDecoder as it is more specialized / not as chatty in generic case.

File: src/DotNetty.Handlers/Tls/TlsHandler.cs (hunk 22 above)
Comment (nayato) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: that's correct

File: src/DotNetty.Handlers/Tls/TlsHandler.cs [hunk 26]
```diff
@@ -280,61 +521,93 @@ public override void Flush(IChannelHandlerContext context)
...
 
-                var buf = (IByteBuffer)msg;
-                buf.ReadBytes(this.sslStream, buf.ReadableBytes); // this leads to FinishWrap being called 0+ times
+                    if (messages.Count == 1)
+                    {
+                        buf = (IByteBuffer)messages[0];
+                    }
+                    else
+                    {
+                        buf = context.Allocator.Buffer((int)this.pendingUnencryptedWrites.CurrentSize);
```
Comment (mtuchkov) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: Why do we need buf? why can't we write to sslStream directly?

**Comment by mastermanu** [comment]
:shipit:

File: src/DotNetty.Handlers/Tls/TlsHandler.cs (hunk 26 above)
Comment (nayato) [inline src/DotNetty.Handlers/Tls/TlsHandler.cs]: not to package every single buffer as a separate SSL record. Every call to `SslStream.Write(...)` results in a new SSL record.

**Pull request merged**